pkg/apis/kubeflow.org/v2alpha1/trainjob_types.go

```diff
@@ -15,3 +15,235 @@ limitations under the License.
+// +kubebuilder:printcolumn:name="Age",type=date,JSONPath=`.metadata.creationTimestamp`
+
+// TrainJob represents configuration of a training job.
+type TrainJob struct {
+	metav1.TypeMeta `json:",inline"`
+
+	// Standard object's metadata.
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	// Specification of the desired TrainJob.
+	Spec TrainJobSpec `json:"spec,omitempty"`
+
+	// Current status of TrainJob.
+	Status TrainJobStatus `json:"status,omitempty"`
+}
+
+// +kubebuilder:object:root=true
+
+// TrainJobList is a collection of training jobs.
+type TrainJobList struct {
+	metav1.TypeMeta `json:",inline"`
+
+	// Standard list metadata.
+	metav1.ListMeta `json:"metadata,omitempty"`
+
+	// List of TrainJobs.
+	Items []TrainJob `json:"items"`
+}
+
+// TrainJobSpec represents specification of the desired TrainJob.
+type TrainJobSpec struct {
+	// Reference to the training runtime.
+	TrainingRuntimeRef TrainingRuntimeRef `json:"trainingRuntimeRef"`
+
+	// Configuration of the desired trainer.
+	Trainer *Trainer `json:"trainer,omitempty"`
+
+	// Configuration of the training dataset.
+	DatasetConfig *DatasetConfig `json:"datasetConfig,omitempty"`
+
+	// Configuration of the pre-trained and trained model.
+	ModelConfig *ModelConfig `json:"modelConfig,omitempty"`
+
+	// Labels to apply for the derivative JobSet and Jobs.
+	// They will be merged with the TrainingRuntime values.
+	Labels map[string]string `json:"labels,omitempty"`
+
+	// Annotations to apply for the derivative JobSet and Jobs.
+	// They will be merged with the TrainingRuntime values.
+	Annotations map[string]string `json:"annotations,omitempty"`
+
+	// Custom overrides for the training runtime.
+	PodSpecOverrides []PodSpecOverrides `json:"podSpecOverrides,omitempty"`
+
+	// Whether the controller should suspend the running TrainJob.
+	// Defaults to false.
+	Suspend *bool `json:"suspend,omitempty"`
+
+	// ManagedBy field indicates the controller that manages a TrainJob.
+	ManagedBy *string `json:"managedBy,omitempty"`
+}
+
+// TrainingRuntimeRef represents the reference to the existing training runtime.
+type TrainingRuntimeRef struct {
+	// Name of the runtime being referenced.
+	// When namespaced-scoped TrainingRuntime is used, the TrainJob must have
+	// the same namespace as the deployed runtime.
+	Name string `json:"name"`
+
+	// APIGroup of the runtime being referenced.
+	// Defaults to `kubeflow.org`.
+	APIGroup *string `json:"apiGroup,omitempty"`
+
+	// Kind of the runtime being referenced.
+	// It must be one of TrainingRuntime or ClusterTrainingRuntime.
+	// Defaults to ClusterTrainingRuntime.
+	Kind *string `json:"kind,omitempty"`
+}
+
+// Trainer represents the desired trainer configuration.
+// Every training runtime contains `trainer` container which represents Trainer.
+type Trainer struct {
+	// Docker image for the training container.
+	Image string `json:"image,omitempty"`
```

Comment:
Could Image be a pointer to the string? The validation can differentiate between unspecified and empty string.

Reply:
@tenzen-y @shravan-achar Do we want to introduce pointer or `+optional` for image since user can skip the image for Trainer?
E.g. Kubernetes for container users `+optional` validation: https://github.com/kubernetes/api/blob/master/core/v1/types.go#L2708C6-L2709

Reply:
As far as I understand, we can skip specifying this image field when the TrainingRuntime has the image.

```go
    // +optional
    Image *string `json:"image,omitempty"`
```

Regarding to the optional vs required, please refer to the above my comment.